federation/pkg/federation-controller/ingress/ingress_controller.go

```diff
@@ -307,71 +652,92 @@ func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
+			clusterIPName, clusterIPNameExists := clusterIngress.ObjectMeta.Annotations[staticIPNameKeyReadonly]
+			baseLBStatusExists := len(baseIngress.Status.LoadBalancer.Ingress) > 0
+			clusterLBStatusExists := len(clusterIngress.Status.LoadBalancer.Ingress) > 0
+			logStr := fmt.Sprintf("Cluster ingress %q has annotation %q=%q, loadbalancer status exists? [%v], federated ingress has annotation %q=%q, loadbalancer status exists? [%v].  %%s annotation and/or loadbalancer status from cluster ingress to federated ingress.", ingress, staticIPNameKeyReadonly, clusterIPName, clusterLBStatusExists, staticIPNameKeyWritable, baseIPName, baseLBStatusExists)
+			if (!baseIPAnnotationExists && clusterIPNameExists) || (!baseLBStatusExists && clusterLBStatusExists) { // copy the IP name from the readonly annotation on the cluster ingress, to the writable annotation on the federated ingress
+				glog.V(4).Infof(logStr, "Transferring")
+				if !baseIPAnnotationExists && clusterIPNameExists {
+					baseIngress.ObjectMeta.Annotations[staticIPNameKeyWritable] = clusterIPName
+				}
+				if !baseLBStatusExists && clusterLBStatusExists {
+					lbstatusObj, lbErr := conversion.NewCloner().DeepCopy(&clusterIngress.Status.LoadBalancer)
+					lbstatus, ok := lbstatusObj.(*v1.LoadBalancerStatus)
+					if lbErr != nil || !ok {
+						glog.Errorf("Internal error: Failed to clone LoadBalancerStatus of %q in cluster %q while attempting to update master loadbalancer ingress status, will try again later. error: %v, Object to be cloned: %v", ingress, cluster.Name, lbErr, lbstatusObj)
+						ic.deliverIngress(ingress, ic.ingressReviewDelay, true)
+						return
 					}
+					baseIngress.Status.LoadBalancer = *lbstatus
+				}
+				glog.V(4).Infof("Attempting to update base federated ingress: %v", baseIngress)
+				if _, err = ic.federatedApiClient.Extensions().Ingresses(baseIngress.Namespace).Update(baseIngress); err != nil {
+					glog.Errorf("Failed to add static IP annotation to federated ingress %q, will try again later: %v", ingress, err)
+					ic.deliverIngress(ingress, ic.ingressReviewDelay, true)
+					return
 				} else {
-					glog.Errorf("Failed to get federated ingress %q: %v", ingress, err)
+					glog.V(4).Infof("Successfully added static IP annotation to federated ingress: %q", ingress)
+					ic.deliverIngress(ingress, ic.smallDelay, false)
+					return
 				}
+			} else {
+				glog.V(4).Infof(logStr, "Not transferring")
 			}
-			// Update existing ingress, if needed.
-			if !util.ObjectMetaIsEquivalent(desiredIngress.ObjectMeta, clusterIngress.ObjectMeta) ||
-				!reflect.DeepEqual(desiredIngress.Spec, clusterIngress.Spec) {
-				// TODO: In some cases Ingress controllers in the clusters add annotations, so we ideally need to exclude those from
-				// the equivalence comparison to cut down on unnecessary updates.
+			// Update existing cluster ingress, if needed.
+			if util.ObjectMetaIsEquivalent(desiredIngress.ObjectMeta, clusterIngress.ObjectMeta) &&
```

Comment:
sigh, there is also ObjectMetaEquivalent: https://github.com/kubernetes/kubernetes/blob/069929bbfd13c4c598c3657a8868888ade9e0b55/federation/pkg/federation-controller/util/meta.go#L40 :)